app/models/post.rb

```diff
@@ -13,6 +13,27 @@ class Post < ApplicationRecord
 
   has_many :likes, as: :likable, dependent: :destroy
 
+  scope :filter_by, lambda { |author, text, published_date|
+    posts = self
+
+    posts = posts.joins(:user).where('users.nickname LIKE ?', "%#{author}%") if author.present?
+
+    posts = posts.where('title LIKE ? OR body LIKE ?', "%#{text}%", "%#{text}%") if text.present?
+
+    if published_date.present?
+      case published_date
+      when 'last_day'
+        posts = posts.where('published_at >= ?', 1.day.ago)
+      when 'last_week'
+        posts = posts.where('published_at >= ?', 1.week.ago)
+      when 'last_month'
+        posts = posts.where('published_at >= ?', 1.month.ago)
+      end
+    end
+
+    posts
+  }
```

Comment:
Thinking about this, I think this could be improved by creating a scope for each type of filter and then a final scope that combines all scopes.

Something like:

```
scope :by_author, -> (author) {
  joins(:user).where('users.nickname LIKE ?', "%#{author}%") if author.present?
}

scope :by_text, -> (text) {
  where('title LIKE ? OR body LIKE ?', "%#{text}%", "%#{text}%") if text.present?
}

scope :by_published_date, -> (published_date) {
  return if published_date.blank?

  case published_date
  when 'last_day'
    where('published_at >= ?', 1.day.ago)
  when 'last_week'
    where('published_at >= ?', 1.week.ago)
  when 'last_month'
    where('published_at >= ?', 1.month.ago)
  end
}

scope :filter_by, -> (author, text, published_date) {
  by_author(author).by_text(text).by_published_date(published_date)
}
```